### PR TITLE
perf: Rewrite retention query to use arrays instead of joins

### PR DESCRIPTION
**Pull request by timgl**
## Problem

Retention queries are slow, mostly because they read the events table 3 times because of the joins. This new version uses arrays so it only has to read that table once, and uses a lot less memory

## Changes

Old query (all events, 10 weeks)
image

New query 
image

👉 _Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review._

### CODE REVIEW

File: posthog/hogql_queries/insights/retention_query_runner.py [hunk 1]
```diff
@@ -98,156 +193,122 @@ def retention_events_query(self, event_query_type) -> ast.SelectQuery:
...
     def to_query(self) -> ast.SelectQuery | ast.SelectUnionQuery:
         placeholders = {
             "actor_query": self.actor_query(),
         }
+
         with self.timings.measure("retention_query"):
             retention_query = parse_select(
                 """
-                    SELECT actor_activity.breakdown_values         AS breakdown_values,
+                    SELECT [actor_activity.breakdown_values]       AS breakdown_values,
```
[the remark this answers is not in the document]
Comment (aspicer) [reply posthog/hogql_queries/insights/retention_query_runner.py:311]: I think this part just makes it the same as the old version.

The new version uses an arrayJoin which returns these as raw elements, but the old version had this line in a subselect that turned it into an array
`[dateDiff('week', toStartOfWeek(toDateTime64('2020-06-07 00:00:00.000000', 6, 'UTC'), 0), toStartOfWeek(toTimeZone(events.timestamp, 'UTC'), 0))] AS breakdown_values`

**Comment by aspicer** [comment]
Ran it locally and did some testing. Queries for recurring events seem like they return the same values as existing queries.

Queries "for the first time" seem like they may not be working quite right:
master:
Screenshot from 2024-05-29 11-04-06

this branch:
Screenshot from 2024-05-29 11-06-58

**Comment by timgl** [comment]
@aspicer Fixed now!

**Review by aspicer — APPROVED** [review]
Thanks for the extra tests.

Tested more locally. Looks good to me!
:rocket:

**Pull request merged**